File: android/app/src/main/kotlin/com/zilpaymobile/HIDDevice.java
package com.zilpaymobile;

import android.hardware.usb.UsbConstants;
import android.hardware.usb.UsbDevice;
import android.hardware.usb.UsbDeviceConnection;
import android.hardware.usb.UsbEndpoint;
import android.hardware.usb.UsbInterface;
import android.hardware.usb.UsbManager;
import android.hardware.usb.UsbRequest;
import android.os.Handler;
import android.os.Looper;

import java.io.ByteArrayOutputStream;
import java.nio.ByteBuffer;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import io.flutter.plugin.common.MethodChannel;

public class HIDDevice {

    private final UsbDeviceConnection connection;
    private final UsbInterface dongleInterface;
    private final UsbEndpoint in;
    private final UsbEndpoint out;
    private final byte[] transferBuffer;
    private final ExecutorService executor;
    private final Handler mainThreadHandler = new Handler(Looper.getMainLooper());


    private static final int HID_BUFFER_SIZE = 64;
    private static final int LEDGER_DEFAULT_CHANNEL = 1;

    public HIDDevice(UsbManager manager, UsbDevice device) throws SecurityException {
        dongleInterface = device.getInterface(0);
        UsbEndpoint endpointIn = null;
        UsbEndpoint endpointOut = null;
        for (int i = 0; i < dongleInterface.getEndpointCount(); i++) {
            UsbEndpoint tmpEndpoint = dongleInterface.getEndpoint(i);
            if (tmpEndpoint.getDirection() == UsbConstants.USB_DIR_IN) {
                endpointIn = tmpEndpoint;
            } else {
                endpointOut = tmpEndpoint;
            }
        }
        in = endpointIn;
        out = endpointOut;
        connection = manager.openDevice(device);
        connection.claimInterface(dongleInterface, true);
        transferBuffer = new byte[HID_BUFFER_SIZE];
        executor = Executors.newSingleThreadExecutor();
    }

    public void exchange(final byte[] commandSource, final MethodChannel.Result result) {
        Runnable exchangeRunnable = () -> {
            try {
                ByteArrayOutputStream response = new ByteArrayOutputStream();
                byte[] responseData;
                int offset = 0;
                byte[] command = LedgerHelper.wrapCommandAPDU(LEDGER_DEFAULT_CHANNEL, commandSource, HID_BUFFER_SIZE);

                UsbRequest requestOut = new UsbRequest();
                if (!requestOut.initialize(connection, out)) {
                    throw new Exception("I/O error on requestOut initialize");
                }
                while (offset != command.length) {
                    int blockSize = Math.min(command.length - offset, HID_BUFFER_SIZE);
                    System.arraycopy(command, offset, transferBuffer, 0, blockSize);
                    if (!requestOut.queue(ByteBuffer.wrap(transferBuffer), HID_BUFFER_SIZE)) {
                        requestOut.close();
                        throw new Exception("I/O error on requestOut queue");
                    }
                    connection.requestWait();
                    offset += blockSize;
                }
                requestOut.close();

                ByteBuffer responseBuffer = ByteBuffer.allocate(HID_BUFFER_SIZE);
                UsbRequest requestIn = new UsbRequest();
                if (!requestIn.initialize(connection, in)) {
                    throw new Exception("I/O error on requestIn initialize");
                }

                while ((responseData = LedgerHelper.unwrapResponseAPDU(LEDGER_DEFAULT_CHANNEL, response.toByteArray(), HID_BUFFER_SIZE)) == null) {
                    responseBuffer.clear();
                    if (!requestIn.queue(responseBuffer, HID_BUFFER_SIZE)) {
                        requestIn.close();
                        throw new Exception("I/O error on requestIn queue");
                    }
                    connection.requestWait();
                    responseBuffer.rewind();
                    responseBuffer.get(transferBuffer, 0, HID_BUFFER_SIZE);
                    response.write(transferBuffer, 0, HID_BUFFER_SIZE);
                }
                requestIn.close();
                final String hexResponse = toHex(responseData);
                mainThreadHandler.post(() -> result.success(hexResponse));

            } catch (Exception e) {
                mainThreadHandler.post(() -> result.error("ExchangeError", e.getMessage(), null));
            }
        };
        this.executor.submit(exchangeRunnable);
    }

    public void close() {
        if (connection != null) {
            connection.releaseInterface(dongleInterface);
            connection.close();
        }
        if (executor != null && !executor.isShutdown()) {
            executor.shutdown();
        }
    }

    public static String toHex(byte[] buffer) {
        StringBuilder result = new StringBuilder();
        for (byte b : buffer) {
            result.append(String.format("%02x", b));
        }
        return result.toString();
    }
}

File: android/app/src/main/kotlin/hid/HidDevice.kt
package hid

import android.hardware.usb.*
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import kotlin.math.min

class HidDevice(manager: UsbManager, device: UsbDevice) {

    private val connection: UsbDeviceConnection
    private val dongleInterface: UsbInterface = device.getInterface(0)
    private val endpointIn: UsbEndpoint
    private val endpointOut: UsbEndpoint
    private val executor = Executors.newSingleThreadExecutor()

    companion object {
        private const val HID_BUFFER_SIZE = 64
        private const val LEDGER_DEFAULT_CHANNEL = 1
        private const val TIMEOUT_MS = 5000L
    }

    init {
        var inEp: UsbEndpoint? = null
        var outEp: UsbEndpoint? = null
        for (i in 0 until dongleInterface.endpointCount) {
            val endpoint = dongleInterface.getEndpoint(i)
            if (endpoint.direction == UsbConstants.USB_DIR_IN) {
                inEp = endpoint
            } else {
                outEp = endpoint
            }
        }
        endpointIn = inEp ?: throw Exception("Input endpoint not found")
        endpointOut = outEp ?: throw Exception("Output endpoint not found")
        connection = manager.openDevice(device) ?: throw Exception("Failed to open device")
        connection.claimInterface(dongleInterface, true)
    }

    fun exchange(command: ByteArray): ByteArray {
        val future = executor.submit<ByteArray> {
            // Wrap command into HID packets
            val wrappedCommand = LedgerHelper.wrapCommandAPDU(LEDGER_DEFAULT_CHANNEL, command, HID_BUFFER_SIZE)

            // Send packets one by one
            val packets = mutableListOf<ByteArray>()
            var offset = 0
            while (offset < wrappedCommand.size) {
                val packet = ByteArray(HID_BUFFER_SIZE)
                val copySize = min(HID_BUFFER_SIZE, wrappedCommand.size - offset)
                System.arraycopy(wrappedCommand, offset, packet, 0, copySize)
                packets.add(packet)
                offset += HID_BUFFER_SIZE
            }

            // Send each packet
            for ((index, packet) in packets.withIndex()) {
                val bytesSent = connection.bulkTransfer(
                    endpointOut,
                    packet,
                    packet.size,
                    TIMEOUT_MS.toInt()
                )

                if (bytesSent != packet.size) {
                    throw DisconnectedDeviceException("Failed to send packet $index: expected ${packet.size}, sent $bytesSent")
                }

                // Small delay between packets for device processing
                if (index < packets.size - 1) {
                    Thread.sleep(10)
                }
            }

            // Read response
            val response = ByteArrayOutputStream()
            var responseData: ByteArray? = null
            val buffer = ByteArray(HID_BUFFER_SIZE)
            var attempts = 0
            val maxAttempts = 100 // Max 5 seconds with 50ms delays

            while (responseData == null && attempts < maxAttempts) {
                val bytesRead = connection.bulkTransfer(
                    endpointIn,
                    buffer,
                    buffer.size,
                    50 // Short timeout for each read
                )

                if (bytesRead > 0) {
                    response.write(buffer, 0, bytesRead)
                    responseData = LedgerHelper.unwrapResponseAPDU(
                        LEDGER_DEFAULT_CHANNEL,
                        response.toByteArray(),
                        HID_BUFFER_SIZE
                    )
                } else if (bytesRead < 0 && response.size() == 0) {
                    // Only throw if we haven't received any data yet
                    Thread.sleep(50)
                    attempts++
                } else {
                    // No more data but we already have some response
                    Thread.sleep(50)
                    attempts++
                }
            }

            if (responseData == null) {
                throw DisconnectedDeviceException("Failed to receive complete response after $attempts attempts")
            }

            responseData
        }

        return future.get(30, TimeUnit.SECONDS)
    }

    fun close() {
        executor.submit {
            connection.releaseInterface(dongleInterface)
            connection.close()
        }.get()
        executor.shutdown()
    }
}